src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -521,182 +608,464 @@ static const uint32_t *find_union_case (const uint32_t * __restrict union_ops, u
+        uint32_t member_id = ops[1];
+        const uint32_t *jeq_ops = ops + DDS_OP_ADR_JSR (insn);
+        dds_stream_write_pl_member (flags & DDS_OP_FLAG_MU, member_id, os, data, jeq_ops);
+        ops += 2;
         break;
       }
-      case DDS_OP_RTS: case DDS_OP_JEQ: {
-        abort ();
+      default:
+        abort (); /* other ops not supported at this point */
         break;
-      }
     }
   }
+  /* write serialized size in dheader */
+  *((uint32_t *) (os->m_buffer + data_offs - 4)) = os->m_index - data_offs;
+  return ops;
 }
 
-static void realloc_sequence_buffer_if_needed (dds_sequence_t * __restrict seq, uint32_t num, uint32_t elem_size, bool init)
+static const uint32_t *dds_stream_write_plLE (dds_ostreamLE_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
 {
-  const uint32_t size = num * elem_size;
+  /* skip PLC op */
+  ops++;
 
-  /* maintain max sequence length (may not have been set by caller) */
-  if (seq->_length > seq->_maximum)
-    seq->_maximum = seq->_length;
+  /* alloc space for dheader */
+  dds_os_put4LE (os, 0xffffffff);
+  uint32_t data_offs = os->x.m_index;
 
-  if (num > seq->_maximum && seq->_release)
+  uint32_t insn;
+  while ((insn = *ops) != DDS_OP_RTS)
   {
-    seq->_buffer = ddsrt_realloc (seq->_buffer, size);
-    if (init)
+    switch (DDS_OP (insn))
+    {
+      case DDS_OP_JEQ: {
+        uint32_t flags = DDS_JEQ_FLAGS (insn);
+        uint32_t member_id = ops[1];
+        const uint32_t *jeq_ops = ops + DDS_OP_ADR_JSR (insn);
+        dds_stream_write_pl_memberLE (flags & DDS_OP_FLAG_MU, member_id, os, data, jeq_ops);
+        ops += 2;
+        break;
+      }
+      default:
+        abort (); /* other ops not supported at this point */
+        break;
+    }
+  }
+  /* write serialized size in dheader */
+  *((uint32_t *) (os->x.m_buffer + data_offs - 4)) = ddsrt_toLE4u (os->x.m_index - data_offs);
+  return ops;
+}
+
+static const uint32_t *dds_stream_write_plBE (dds_ostreamBE_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+{
+  /* skip PLC op */
+  ops++;
+
+  /* alloc space for dheader */
+  dds_os_put4BE (os, 0xffffffff);
+  uint32_t data_offs = os->x.m_index;
+
+  uint32_t insn;
+  while ((insn = *ops) != DDS_OP_RTS)
+  {
+    switch (DDS_OP (insn))
+    {
+      case DDS_OP_JEQ: {
+        uint32_t flags = DDS_JEQ_FLAGS (insn);
+        uint32_t member_id = ops[1];
+        const uint32_t *jeq_ops = ops + DDS_OP_ADR_JSR (insn);
+        dds_stream_write_pl_memberBE (flags & DDS_OP_FLAG_MU, member_id, os, data, jeq_ops);
+        ops += 2;
+        break;
+      }
+      default:
+        abort (); /* other ops not supported at this point */
+        break;
+    }
+  }
+  /* write serialized size in dheader */
+  *((uint32_t *) (os->x.m_buffer + data_offs - 4)) = ddsrt_toBE4u (os->x.m_index - data_offs);
+  return ops;
+}
+
+// Native endianness
+#define NAME_BO_EXT
```

Comment:
It would probably be better to only do LE and BE, and then simply map the native endianness one onto the correct one. (That would then also be true for some of the functions like `dds_stream_write_pl`.)

Reply:
You're right, that saves a lot of code! Fixed in 1b3eb65